### PR TITLE
Miri: give machine the chance to tag all allocations

### PR DESCRIPTION
**Pull request by RalfJung**
r? @oli-obk

The Miri side of this is at https://github.com/rust-lang/rust/pull/61278.

### CODE REVIEW

File: src/librustc/mir/interpret/value.rs [hunk 1]
```diff
@@ -140,23 +140,18 @@ impl<'tcx> Scalar<()> {
...
-        where Tag: Default
-    {
-        self.with_tag(Tag::default())
-    }
 }
 
 impl<'tcx, Tag> Scalar<Tag> {
     #[inline]
     pub fn erase_tag(self) -> Scalar {
+        // Used by error reporting code to avoid having the error type depend on `Tag`
```
Comment (oli-obk) [inline src/librustc/mir/interpret/value.rs]: move this up to a `Doc` comment on the function?

File: src/librustc/mir/interpret/value.rs [hunk 2]
```diff
@@ -476,27 +471,11 @@ impl<Tag> fmt::Display for ScalarMaybeUndef<Tag> {
...
-    {
-        self.with_tag(Tag::default())
-    }
-}
-
 impl<'tcx, Tag> ScalarMaybeUndef<Tag> {
     #[inline]
     pub fn erase_tag(self) -> ScalarMaybeUndef
     {
+        // Used by error reporting code to avoid having the error type depend on `Tag`
```
Comment (oli-obk) [inline src/librustc/mir/interpret/value.rs]: same here

File: src/librustc_mir/interpret/eval_context.rs [hunk 3]
```diff
@@ -43,7 +43,10 @@ pub struct InterpretCx<'a, 'mir, 'tcx: 'a + 'mir, M: Machine<'a, 'mir, 'tcx>> {
     pub(crate) stack: Vec<Frame<'mir, 'tcx, M::PointerTag, M::FrameExtra>>,
 
     /// A cache for deduplicating vtables
-    pub(super) vtables: FxHashMap<(Ty<'tcx>, Option<ty::PolyExistentialTraitRef<'tcx>>), AllocId>,
+    pub(super) vtables: FxHashMap<
+            (Ty<'tcx>, Option<ty::PolyExistentialTraitRef<'tcx>>),
+            Pointer<M::PointerTag>
+        >,
```
Comment (oli-obk) [inline src/librustc_mir/interpret/eval_context.rs]: rustfmt says one indendation level less: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=29708e9888f92aff5cd673be6f9b69d0

**Comment by RalfJung** [comment]
I addressed all comments.

**Review by oli-obk — COMMENTED** [review]
oops, sorry, github tricked me again. I didn't send these comments

**Comment by oli-obk** [comment]
r=me with them resolved

File: src/librustc_mir/interpret/operand.rs [hunk 4]
```diff
@@ -537,44 +506,53 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> InterpretCx<'a, 'mir, 'tcx, M>
...
         let layout = from_known_layout(layout, || {
             self.layout_of(self.monomorphize(val.ty)?)
         })?;
-        Ok(OpTy {
-            op,
-            layout,
-        })
+        let op = match val.val {
+            ConstValue::ByRef(ptr, _alloc) => {
+                // We rely on mutability being set correctly in that allocation to prevent writes
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/librustc_mir/interpret/operand.rs]: Not sure what you mean?

*If* the machine does enough extra checks, then the check that "you can't write through a shared reference" is indeed enough (I think). But that's a big if.

[the remark this answers is not in the document]
Comment (oli-obk) [reply src/librustc_mir/interpret/operand.rs]: we used to take the alignment from the allocation via `alloc.align`, but now you take it from the `layout`. So the comment about us relying on the .... oh wait XD the comment is about mutability not alignment, ignore me.

[the remark this answers is not in the document]
Comment (RalfJung) [reply src/librustc_mir/interpret/operand.rs]: The next prat of the comment about "copy on demand" is somewhat wrong though, Miri actually copies even read-only statics on first read.

File: src/librustc_mir/interpret/memory.rs [hunk 5]
```diff
@@ -305,53 +317,70 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> Memory<'a, 'mir, 'tcx, M> {
...
+                            // no need to report anything, the const_eval call takes care of that
+                            // for statics
+                            assert!(tcx.is_static(def_id));
+                            match err {
+                                ErrorHandled::Reported => InterpError::ReferencedConstant,
+                                ErrorHandled::TooGeneric => InterpError::TooGeneric,
+                            }
+                        })?;
+                    // Make sure we use the ID of the resolved memory, not the lazy one!
+                    let id = raw_const.alloc_id;
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/librustc_mir/interpret/memory.rs]: Yeah, that should work I think. Not sure about the interactions with... everything else though.^^

File: src/librustc_mir/interpret/operand.rs [hunk 6]
```diff
@@ -537,44 +506,55 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> InterpretCx<'a, 'mir, 'tcx, M>
...
+            },
+            ConstValue::Scalar(x) =>
+                Operand::Immediate(Immediate::Scalar(tag_scalar(x).into())),
+            ConstValue::Slice { data, start, end } => {
+                // We rely on mutability being set correctly in `data` to prevent writes
+                // where none should happen.
+                let ptr = Pointer::new(
+                    self.tcx.alloc_map.lock().allocate(data),
+                    Size::from_bytes(start as u64), // offset: `start`
+                );
```
Comment (RalfJung) [inline src/librustc_mir/interpret/operand.rs]: This code just struck me as odd (I just moved it around). Doesn't this mean we use up tons of `AllocId` when looking at the same constant many times?

File: src/librustc_mir/interpret/memory.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/librustc_mir/interpret/memory.rs]: Opened a new issue to track this: https://github.com/rust-lang/rust/issues/61345

File: src/librustc_mir/interpret/operand.rs (hunk 6 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/operand.rs]: yes. Initially I thought that would be necessary, but I think this code can only ever occur if the constant is consumed (copied) and not ever have its address taken. Will need some investigation, but we should be able to resolve it together with #61345

**Comment by RalfJung** [comment]
Rebased.

@bors r=oli-obk

**Comment by RalfJung** [comment]
Looks spurious.

@bors retry

**Comment by RalfJung** [comment]
Cc https://github.com/rust-lang/rust/issues/55861 for the CI failure.

**Comment by RalfJung** [comment]
Uh, wrong command.

@bors r=oli-obk

**Comment by RalfJung** [comment]
On the critical path to get Miri working again, and has been bitrotting a lot.

@bors p=1

**Comment by RalfJung** [comment]
Timeout on the macOS builder?

@bors retry

**Comment by RalfJung** [comment]
Another macOS timeout? Cc @rust-lang/infra

@bors retry

**Pull request merged**